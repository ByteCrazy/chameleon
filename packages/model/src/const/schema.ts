export enum CNodePropsTypeEnum {
  SLOT = 'SLOT',
  FUNCTION = 'FUNCTION',
  EXPRESSION = 'EXPRESSION',
}

export enum ExportTypeEnum {
  DESIGN = 'design',
  SAVE = 'save',
}

export type ExportType = ExportTypeEnum | `${ExportTypeEnum}`;

export enum SlotRenderType {
  FUNC = 'FUNC',
  COMP = 'COMP',
}
