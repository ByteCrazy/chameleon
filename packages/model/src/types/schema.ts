import {
  array,
  assign,
  literal,
  object,
  omit,
  optional,
  string,
} from 'superstruct';
import { CNodePropsTypeEnum } from '../const/schema';
import {
  CNodeDataStructDescribe,
  CNodeDataType,
  FunctionPropType,
} from './node';

export enum InnerComponentNameEnum {
  PAGE = 'CPage',
  COMPONENT = 'CComponent',
}

export type CSchemaDataType = CNodeDataType & {
  componentName: InnerComponentNameEnum.PAGE | `${InnerComponentNameEnum.PAGE}`;
  beforeRender?: FunctionPropType;
  afterRender?: FunctionPropType;
  methods?: FunctionPropType[];
};

export const FunctionPropertyTypeDescribe = object({
  type: literal(CNodePropsTypeEnum.FUNCTION),
  value: string(),
});

export const CSchemaDataTypeDescribe = assign(
  omit(CNodeDataStructDescribe, ['componentName']),
  object({
    componentName: literal(InnerComponentNameEnum.PAGE),
    beforeRender: optional(FunctionPropertyTypeDescribe),
    afterRender: optional(FunctionPropertyTypeDescribe),
    methods: optional(array(FunctionPropertyTypeDescribe)),
  })
);
