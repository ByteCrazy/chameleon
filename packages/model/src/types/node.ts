import {
  string,
  number,
  boolean,
  any,
  union,
  object,
  literal,
  optional,
  record,
  dynamic,
  array,
} from 'superstruct';
import { CNodePropsTypeEnum } from '../const/schema';
import { ThirdLibType } from './base';

export type NormalPropType = string | boolean | number | Record<string, any>;

export type RenderPropType = {
  type: CNodePropsTypeEnum.SLOT;
  value: CNodeDataType[];
};

export type JSExpressionPropType = {
  type: CNodePropsTypeEnum.EXPRESSION;
  value: string;
};

export type FunctionPropType = {
  type: CNodePropsTypeEnum.FUNCTION;
  value: string;
};

export type SpecialProp = RenderPropType | JSExpressionPropType;

export type PropType = NormalPropType | SpecialProp;

export const PropsDataStructDescribe: any = union([
  string(),
  number(),
  boolean(),
  object({
    type: literal(CNodePropsTypeEnum.SLOT),
    // here can't use PropsDataStructDescribe, it will  caused  "Maximum call stack size exceeded" error
    value: dynamic(() => {
      return CNodeDataStructDescribe;
    }),
  }),
  object({
    type: literal(CNodePropsTypeEnum.EXPRESSION),
    value: string(),
  }),
  object({
    type: literal(CNodePropsTypeEnum.FUNCTION),
    value: string(),
  }),
]);

// 开发模式使用的 key,导出为生产模式时，需要移除
export const DevKey = ['configure'];

export type CNodeDataType = {
  id: string;
  componentName: string;
  props?: Record<string, PropType>;
  children?: CNodeDataType[];
  /**
   * only used in dev mode, if you are run in prod, this key will be undefined
   *
   * @type {Record<any, any>}
   */
  configure?: Record<any, any>;
  style?: string;
  className?: string;
  ref?: string;
  onEvents?: Record<
    string,
    {
      actions?: {
        // 需要场景化，暂时 string ，需要嘿编辑器 行为绑定
        actionType: string;
        options?: Record<any, any>;
      }[];
    }
  >;
};

export const CNodeDataStructDescribe: any = object({
  id: optional(string()),
  componentName: string(),
  props: optional(record(string(), PropsDataStructDescribe)),
  children: dynamic(() => {
    return optional(array(CNodeDataStructDescribe));
  }),
  configure: optional(any()),
  style: optional(string()),
  className: optional(string()),
  ref: optional(string()),
});
