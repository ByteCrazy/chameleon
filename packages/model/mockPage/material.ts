import { AdvanceDataType, CMaterialType } from '../src/types/material';

export const mockMaterialData: CMaterialType[] = [
  {
    title: '测试物料组件',
    componentName: 'Header',
    category: '默认',
    groupName: 'A',
    npm: {
      package: 'antd',
      version: '1.0.0',
      exportName: 'Button',
      destructuring: true,
    },
    props: [
      {
        name: 'text',
        title: '文本',
        defaultValue: '按钮',
        setters: ['StringSetter'],
        valueType: {
          type: AdvanceDataType.SHAPE,
          value: [
            {
              name: 'key1',
              title: 'key1',
              valueType: {
                type: AdvanceDataType.SHAPE,
                value: [
                  {
                    name: 'key2',
                    title: 'key2',
                    valueType: BaseDataType.STRING,
                  },
                ],
              },
            },
            {
              name: 'renderItem',
              title: '渲染子元素',
              valueType: SpecialDataType.COMPONENT,
            },
            {
              name: 'expression',
              title: '表达式',
              valueType: SpecialDataType.EXPRESSION,
            },
          ],
        } as ShapeDataType,
      },
      {
        name: 'renderItem',
        title: '渲染子元素',
        valueType: SpecialDataType.COMPONENT,
      },
      {
        name: 'expression',
        title: '表达式',
        valueType: SpecialDataType.EXPRESSION,
      },
    ],
    snippets: [
      {
        title: 'demo',
        schema: {
          props: {
            a: 1,
          },
        },
      },
      {
        title: 'demo2',
        schema: {
          props: {
            a: 1,
          },
        },
      },
    ],
  },
  {
    title: '测试物料组件',
    componentName: 'Header2',
    category: '高级2',
    groupName: 'B',
    npm: {
      package: 'antd',
      version: '1.0.0',
      exportName: 'Button',
      destructuring: true,
    },
    props: [
      {
        name: 'text',
        title: '文本',
        defaultValue: '按钮',
        setters: ['StringSetter'],
        valueType: {
          type: AdvanceDataType.SHAPE,
          value: [
            {
              name: 'key1',
              title: 'key1',
              valueType: {
                type: AdvanceDataType.SHAPE,
                value: [
                  {
                    name: 'key2',
                    title: 'key2',
                    valueType: BaseDataType.STRING,
                  },
                ],
              },
            },
            {
              name: 'renderItem',
              title: '渲染子元素',
              valueType: SpecialDataType.COMPONENT,
            },
            {
              name: 'expression',
              title: '表达式',
              valueType: SpecialDataType.EXPRESSION,
            },
          ],
        } as ShapeDataType,
      },
      {
        name: 'renderItem',
        title: '渲染子元素',
        valueType: SpecialDataType.COMPONENT,
      },
      {
        name: 'expression',
        title: '表达式',
        valueType: SpecialDataType.EXPRESSION,
      },
    ],
    snippets: [
      {
        title: 'demo',
        schema: {
          props: {
            a: 1,
          },
        },
      },
      {
        title: 'demo2',
        schema: {
          props: {
            a: 1,
          },
        },
      },
    ],
  },
];
