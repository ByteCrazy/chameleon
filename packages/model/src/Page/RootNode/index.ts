import { merge, omit } from 'lodash-es';
import { CPage } from '..';
import { ExportType, ExportTypeEnum } from '../../const/schema';
import { CMaterials } from '../../Material';
import { CRootNodeDataType, CRootNodeDataTypeDescribe, InnerComponentNameEnum } from '../../types/rootNode';
import { clearSchema, getNode, getRandomStr } from '../../util';
import { checkComplexData } from '../../util/dataCheck';
import { isArray, isPlainObject } from '../../util/lodash';
import { DataModelEmitter, DataModelEventType } from '../../util/modelEmitter';
import { CNode } from './Node/index';
import { CProp } from './Node/prop';

export type CRootNodeModelDataType = Omit<CRootNodeDataType, 'children'> & {
  id: string;
  children: CNode[];
  props: Record<string, CProp>;
  configure: Required<CRootNodeDataType>['configure'];
};

export const checkRootNode = (data: any): CRootNodeDataType => {
  checkComplexData({
    data: data,
    dataStruct: CRootNodeDataTypeDescribe,
    throwError: false,
  });
  return data;
};

export const parseSchema = (
  data: CRootNodeDataType | CRootNodeModelDataType,
  parent: CRootNode,
  materials: CMaterials
): CRootNodeModelDataType => {
  const res: CRootNodeModelDataType = {
    ...data,
    id: getRandomStr(),
    props: {} as any,
    componentName: InnerComponentNameEnum.ROOT_CONTAINER,
    children: [],
    methods: data.methods || [],
    configure: merge(data.configure || {}, {
      propsSetter: {},
      advanceSetter: {},
    }),
  };
  let child: any = [];
  if (isArray(data.children)) {
    child = data.children.map((el: any) => {
      if (el instanceof CNode) {
        return el;
      }
      if (isPlainObject(el)) {
        return new CNode(el, { parent: parent, materials });
      } else {
        return el;
      }
    });
  } else {
    if ((data.children as unknown) instanceof CNode) {
      child.push(data.children);
    } else {
      if (data.children && isPlainObject(data.children)) {
        child.push(new CNode(data.children, { parent: parent, materials }));
      }
    }
  }

  const propsKeys = Object.keys(data.props || {});

  if (propsKeys.length) {
    propsKeys.forEach((propKey) => {
      const targetProps = data.props?.[propKey];
      if (targetProps instanceof CProp) {
        res.props[propKey] = targetProps;
      } else {
        res.props[propKey] = new CProp(propKey, targetProps || '', {
          parent: parent,
          materials,
        });
      }
    });
  }
  res.children = child;

  return res;
};

type OnNodeChangeType = (params: DataModelEventType['onNodeChange']) => void;

export class CRootNode {
  private rawData: CRootNodeDataType;
  private data: CRootNodeModelDataType;
  nodeType = InnerComponentNameEnum.ROOT_CONTAINER;
  emitter = DataModelEmitter;
  materialsModel: CMaterials;
  listenerHandle: (() => void)[];
  onChangeCbQueue: OnNodeChangeType[];
  parent: CPage | null;
  constructor(data: any, { parent, materials }: { parent: CPage | null; materials: CMaterials }) {
    this.materialsModel = materials;
    this.rawData = JSON.parse(JSON.stringify(data));
    this.data = parseSchema(data, this, materials);
    this.listenerHandle = [];
    this.onChangeCbQueue = [];
    this.registerListener();
    this.parent = parent;
  }

  registerListener() {
    const onNodeChange = (params: DataModelEventType['onNodeChange']) => {
      const { node } = params;
      if (node === this && node.id === this.id) {
        this.onChangeCbQueue.forEach((it) => it(params));
      }
    };
    this.emitter.on('onNodeChange', onNodeChange);
    this.listenerHandle.push(() => {
      this.emitter.off('onNodeChange', onNodeChange);
    });
  }

  onChange(cb: OnNodeChangeType) {
    this.onChangeCbQueue.push(cb);
    return () => {
      this.onChangeCbQueue = this.onChangeCbQueue.filter((el) => el !== cb);
    };
  }

  get id() {
    return this.data.id;
  }

  get value() {
    return this.data;
  }

  get props() {
    return this.data.props;
  }

  get material() {
    const materialModel = this.materialsModel;
    return materialModel?.findByComponentName(this.data.componentName);
  }

  updateWithPlainObj(val?: Partial<CRootNodeModelDataType | CRootNodeDataType>) {
    const newVal: any = {
      ...this.data,
      ...val,
    };
    this.data = parseSchema(newVal, this, this.materialsModel);
    return newVal;
  }

  updateValue(val?: Partial<CRootNodeModelDataType | CRootNodeDataType>) {
    const oldData = this.data;
    this.updateWithPlainObj(val);
    this.emitter.emit('onNodeChange', {
      value: this.data,
      preValue: oldData,
      node: this,
    });
  }

  contains(nodeId: string) {
    const res = getNode(this, nodeId);
    return res;
  }

  /**
   * 从物料 和 node config 中获取合并后的属性，node 中的配置优先级更高
   * @param key 目前只支持 isContainer 属性
   * @returns
   */
  getMaterialConfig(key: 'isContainer') {
    if (this.data.configure[key] !== undefined) {
      return this.data.configure.isContainer;
    } else {
      return this.material?.value.isContainer;
    }
  }

  isContainer() {
    return this.getMaterialConfig('isContainer');
  }

  export(mode: ExportType = ExportTypeEnum.SAVE): CRootNodeDataType {
    const data = this.data;
    const props: any = {};
    Object.keys(data.props || {}).forEach((key) => {
      props[key] = data.props[key].export(mode);
    });
    const children: any[] =
      data.children?.map((child) => {
        return child?.export?.(mode);
      }) || [];

    const tempData: CRootNodeDataType = {
      ...data,
      props: props,
      children: children.filter((el) => el),
    };
    let finalRes: any = omit(tempData, ['id']);
    finalRes = clearSchema(finalRes);

    return finalRes;
  }

  getPlainProps() {
    const data = this.data;
    const props: any = {};
    Object.keys(data.props || {}).forEach((key) => {
      props[key] = data.props[key].export('design');
    });
    return props;
  }

  destroy() {
    this.listenerHandle.forEach((it) => it());
  }

  clone(id?: string) {
    const newData = {
      ...this.export('design'),
      id: id || getRandomStr(),
    };
    return new CRootNode(newData, {
      materials: this.materialsModel,
      parent: null,
    });
  }
}
