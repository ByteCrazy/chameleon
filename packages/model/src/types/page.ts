import {
  any,
  array,
  assign,
  literal,
  object,
  optional,
  string,
  union,
} from 'superstruct';
import {
  LibMetaType,
  ThirdLibTypeDescribe,
  LibMetaTypeDescribe,
  AssetPackage,
  CSSType,
} from './base';
import { FunctionPropType } from './node';
import {
  CRootNodeDataType,
  CRootNodeDataTypeDescribe,
  FunctionPropertyTypeDescribe,
} from './rootNode';

export type ComponentMetaType = {
  componentName: string;
} & LibMetaType;

export enum RenderType {
  PAGE = 'PAGE',
  COMPONENT = 'COMPONENT',
}

export type CPageDataType = {
  version: string;
  name: string;
  css?: CSSType[];
  renderType?: RenderType;
  methods?: FunctionPropType[];
  componentsMeta: ComponentMetaType[];
  thirdLibs?: LibMetaType[];
  componentsTree: CRootNodeDataType;
  // runtime render need
  assets?: AssetPackage[];
};

export const CPageDataTypeDescribe = object({
  version: string(),
  name: string(),
  css: optional(string()),
  renderType: optional(
    union([literal(RenderType.COMPONENT), literal(RenderType.PAGE)])
  ),
  methods: optional(array(FunctionPropertyTypeDescribe)),
  componentsMeta: array(
    assign(
      object({
        componentName: string(),
      }),
      LibMetaTypeDescribe
    )
  ),
  thirdLibs: optional(ThirdLibTypeDescribe),
  componentsTree: CRootNodeDataTypeDescribe,
  assets: optional(array(any())),
});
