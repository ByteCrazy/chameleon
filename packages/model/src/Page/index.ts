import { checkComplexData } from '../util/dataCheck';
import { ComponentMetaType, CPageDataType, CPageDataTypeDescribe } from '../types/page';
import { DataModelEmitter } from '../util/modelEmitter';
import { CRootNode } from './RootNode';
import { ExportType, ExportTypeEnum } from '../const/schema';
import { CMaterials } from '../Material';
import { CNode } from './RootNode/Node';
import { CNodeDataType } from '../types/node';
import { cloneDeep, get, isPlainObject, omit, unionBy } from 'lodash-es';
import { CProp } from './RootNode/Node/prop';
import { CSlot } from './RootNode/Node/slot';
import { clearSchema, findNode, getNode, getRandomStr } from '../util';
import { InnerComponentNameEnum } from '../types/rootNode';
import { AssetPackage } from '../types/base';
import { CMaterialType } from '../types/material';

export const checkPage = (data: any): CPageDataType => {
  checkComplexData({
    data: data,
    dataStruct: CPageDataTypeDescribe,
    throwError: false,
  });

  return data;
};

export const parsePage = (data: CPageDataType, parent: CPage, materials: CMaterials) => {
  return {
    ...data,
    componentsTree: new CRootNode(data.componentsTree, {
      parent: parent,
      materials,
    }),
  };
};

export type CPpageDataModelType = Omit<CPageDataType, 'componentsTree'> & {
  componentsTree: CRootNode;
};

export type PosObj = {
  type: 'CHILD';
  index: number;
  pos: 'BEFORE' | 'AFTER';
};

export type InsertNodePosType = 'BEFORE' | 'AFTER' | 'CHILD_START' | 'CHILD_END' | PosObj;

export class CPage {
  nodeType = 'PAGE';
  rawData: CPageDataType;
  emitter = DataModelEmitter;
  data: CPpageDataModelType;
  parent: null | undefined;
  materialsModel: CMaterials;
  assetPackagesList: AssetPackage[];

  constructor(
    data: CPageDataType,
    options?: {
      materials?: CMaterialType[];
      assetPackagesList?: AssetPackage[];
    }
  ) {
    checkPage(data);
    this.assetPackagesList = options?.assetPackagesList || [];
    this.rawData = JSON.parse(JSON.stringify(data));
    this.materialsModel = new CMaterials(options?.materials || []);
    this.data = parsePage(data, this, this.materialsModel);
  }

  updatePage(data: CPageDataType) {
    const oldData = this.data;
    this.rawData = JSON.parse(JSON.stringify(data));
    this.data = parsePage(data, this, this.materialsModel);
    this.emitter.emit('onPageChange', {
      value: this.data,
      preValue: oldData,
      node: this,
    });
  }

  reloadPage(data: CPageDataType) {
    const oldData = this.data;
    this.rawData = JSON.parse(JSON.stringify(data));
    this.data = parsePage(data, this, this.materialsModel);
    this.emitter.emit('onReloadPage', {
      value: this.data,
      preValue: oldData,
      node: this,
    });
  }

  get value() {
    return this.data;
  }

  // moveNode(from, to, pos) {}
  getNode(id: string) {
    const nodeTree = this.data.componentsTree;
    return getNode(nodeTree, id);
  }

  addNode(newNode: CNode, targetNode: CNode | CRootNode, pos: InsertNodePosType = 'AFTER') {
    if (pos === 'AFTER' || pos === 'BEFORE') {
      const parentNode = targetNode.parent;
      // 说明是容器节点, 只能插入 child
      if (parentNode === null && targetNode instanceof CRootNode) {
        console.warn('Not found parent node');
        return false;
      }

      if (parentNode instanceof CProp) {
        console.warn('CProp can not add node');
        return false;
      }
      // TODO:
      if (parentNode instanceof CSlot) {
        const parentList = parentNode.value.value;
        // find it on children;
        const targetIndex = parentList.findIndex((el) => el === targetNode) ?? -1;
        if (targetIndex >= 0) {
          if (pos === 'BEFORE') {
            parentList.splice(targetIndex, 0, newNode);
          } else {
            parentList.splice(targetIndex + 1, 0, newNode);
          }
          newNode.parent = parentNode;
          parentNode.parent?.updateValue();
          return true;
        }
        return false;
      }
      // TODO:
      if (parentNode instanceof CPage) {
        return false;
      }
      // find it on children;
      const targetIndex = parentNode?.value.children.findIndex((el) => el === targetNode) ?? -1;
      if (targetIndex >= 0) {
        if (pos === 'BEFORE') {
          parentNode?.value.children.splice(targetIndex, 0, newNode);
        } else {
          parentNode?.value.children.splice(targetIndex + 1, 0, newNode);
        }
        newNode.parent = parentNode;
        parentNode?.updateValue();
        return true;
      }
      console.warn('Not found target node');
      return false;
    }

    if (pos === 'CHILD_START') {
      targetNode.value.children.unshift(newNode);
      newNode.parent = targetNode;
      targetNode.updateValue();
      return true;
    }

    if (pos === 'CHILD_END') {
      targetNode.value.children.push(newNode);
      newNode.parent = targetNode;
      targetNode.updateValue();
      return true;
    }

    if (isPlainObject(pos)) {
      const posObj = pos as PosObj;
      if (posObj.type === 'CHILD') {
        const subPos = posObj.pos;
        const index = posObj.index || 0;
        if (subPos === 'BEFORE') {
          targetNode?.value.children.splice(index, 0, newNode);
        } else {
          targetNode?.value.children.splice(index + 1, 0, newNode);
        }
        newNode.parent = targetNode;
        targetNode.updateValue();
        return true;
      } else {
        console.warn('Can not parse pos obj');
      }
    }
    return false;
  }

  createNode(nodeData: CNodeDataType) {
    delete nodeData.id;
    const newNode = new CNode(nodeData, {
      parent: null,
      materials: this.materialsModel,
    });
    return newNode;
  }

  addNodeById(newNode: CNode, targetNodeId: string, pos: InsertNodePosType = 'AFTER') {
    const targetNode = this.getNode(targetNodeId);
    if (targetNode) {
      return this.addNode(newNode, targetNode, pos);
    } else {
      console.warn(`Not find a node by ${targetNodeId}, pls check it`);
      return false;
    }
  }

  copyNode(node: CNode) {
    const newNodeData = node.export('design');
    newNodeData.id = getRandomStr();
    const newNode = new CNode(newNodeData, {
      parent: node.parent,
      materials: this.materialsModel,
    });
    this.addNode(newNode, node, 'AFTER');
    return newNode;
  }

  copyNodeById(nodeId: string) {
    const node = this.getNode(nodeId);
    if (node && node instanceof CNode) {
      return this.copyNode(node);
    } else {
      return false;
    }
  }

  moveNode(from: CNode, to: CNode, pos: InsertNodePosType) {
    this.deleteNode(from);
    let parent: any = to;
    const tempTypeList: InsertNodePosType[] = ['AFTER', 'BEFORE'];
    if (tempTypeList.includes(pos)) {
      parent = to.parent;
    }
    from.parent = parent;

    return this.addNode(from, to, pos);
  }

  moveNodeById(fromId: string, toId: string, pos: InsertNodePosType) {
    const from = this.getNode(fromId);
    const to = this.getNode(toId);
    if (from && to && from instanceof CNode && to instanceof CNode) {
      return this.moveNode(from, to, pos);
    }

    return false;
  }

  // replaceNode(targetNode, node) {}

  deleteNode(node: CNode | CRootNode) {
    const parent = node.parent;
    if (!parent) {
      throw new Error('parent node is null or undefined, pls check it');
    }

    if (parent instanceof CSlot) {
      const childList = parent.value.value;
      const targetIndex = childList.findIndex((el) => el === node);
      const deleteNode = childList[targetIndex];

      childList.splice(targetIndex, 1);
      parent.parent?.updateValue();

      return deleteNode;
    }

    if (parent instanceof CNode || parent instanceof CRootNode) {
      const childList = parent.value.children;
      const targetIndex = childList.findIndex((el) => el === node);
      const deleteNode = childList[targetIndex];

      childList.splice(targetIndex, 1);
      parent.updateValue();
      return deleteNode;
    }
  }

  deleteNodeById(nodeId: string) {
    const node = this.getNode(nodeId);
    if (node) {
      return this.deleteNode(node);
    }
  }

  // TODO
  export(mode: ExportType = ExportTypeEnum.SAVE): CPageDataType {
    const componentsTree = this.data.componentsTree.export(mode);

    const assetPackagesList = this.assetPackagesList;

    const finalAssets: AssetPackage[] = [];
    const componentsMetaList: ComponentMetaType[] = this.materialsModel.usedMaterials.map((it) => {
      const asset = assetPackagesList.find((el) => {
        return el.package === it.value.npm?.package;
      });
      if (asset) {
        finalAssets.push(asset);
      }
      return {
        componentName: it.componentName,
        ...cloneDeep(it.value.npm || {}),
      } as ComponentMetaType;
    });
    // 剔除不合法的meta
    const finalComponentsMetaList = componentsMetaList.filter((el) => {
      if (el.componentName && el.package && el.version) {
        return true;
      }
      return false;
    });
    this.materialsModel.usedMaterials = [];
    let res: CPageDataType = {
      ...this.data,
      componentsTree: clearSchema(componentsTree),
      componentsMeta: finalComponentsMetaList,
      thirdLibs: this.data.thirdLibs,
      assets: [],
    };

    this.data.thirdLibs?.forEach((thirdEl) => {
      const asset = assetPackagesList.find((el) => {
        return thirdEl.package === el.package;
      });
      if (asset) {
        finalAssets.push(asset);
      }
    });

    res.assets = unionBy(finalAssets, (el) => el.package);
    res = omit(res, ['id']) as any;
    return JSON.parse(JSON.stringify(res));
  }

  getRootNode() {
    const nodeTree = this.data.componentsTree;
    return findNode(nodeTree, (item) => {
      const componentName = get(item, 'data.componentName') || get(item, 'rawData.componentName');
      return componentName === InnerComponentNameEnum.ROOT_CONTAINER;
    });
  }
}

export const EmptyPage: CPageDataType = {
  version: '1.0.0',
  name: 'EmptyPage',
  componentsMeta: [],
  componentsTree: {
    componentName: InnerComponentNameEnum.ROOT_CONTAINER,
    props: {},
    children: [],
  },
};
