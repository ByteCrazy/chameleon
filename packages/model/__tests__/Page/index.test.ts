import { CPage } from '../../src/Page/index';
import { CMaterialType } from '../../src/types/material';
import { CNodeDataType } from '../../src/types/node';
import { CPageDataType } from '../../src/types/page';
import { BasePage } from '@chameleon/demo-page';
import { InnerComponentNameEnum } from '../../src/types/schema';

const mockMaterial: CMaterialType[] = [
  {
    title: '按钮',
    componentName: 'Button',
    props: [
      {
        title: '文案',
        name: 'text',
        valueType: 'string',
      },
    ],
    npm: {
      package: 'demo',
      version: '1.0.0',
      exportName: 'Button',
    },
    snippets: [
      {
        title: '测试按钮',
        schema: {
          componentName: 'Button',
        },
      },
    ],
  },
  {
    title: '页首',
    componentName: 'Header',
    props: [
      {
        name: 'key',
        title: '唯一标记',
        valueType: 'string',
      },
      {
        name: 'age',
        title: '年龄',
        valueType: 'number',
      },
      {
        name: 'itemRender',
        title: 'itemRender',
        valueType: 'string',
      },
    ],
    npm: {
      package: 'demo',
      version: '1.0.0',
      exportName: 'Header',
    },
    snippets: [],
  },
];

const mockNodeData: CNodeDataType = {
  id: '1',
  componentName: 'Header',
  props: {
    key: '1',
    age: 12,
    itemRender: {
      type: 'SLOT',
      renderType: 'COMP',
      value: [
        {
          id: '2',
          componentName: 'Button',
        },
      ],
    },
  },
};

const mockPageData: CPageDataType = {
  version: '1.1.0',
  pageName: 'testPage',
  style: '',
  componentsMeta: [
    {
      componentName: 'Button',
      package: 'antd',
      version: '1.0',
      exportName: 'Button',
    },
  ],
  componentsTree: {
    componentName: InnerComponentNameEnum.PAGE,
    children: [mockNodeData],
  },
};

describe('test page model', () => {
  it('new a page instance with jsslot', () => {
    const node = new CPage(BasePage as any);
    expect(node).not.toBeNull();
  });

  it('new a page instance', () => {
    const node = new CPage(mockPageData);
    expect(node).not.toBeNull();
  });

  it('validate a bad page', () => {
    const newPageData = {
      ...mockPageData,
    };
    newPageData.version = 1 as any;
    let isError = false;
    try {
      new CPage(newPageData);
    } catch (e) {
      isError = true;
    }

    expect(isError).toBeTruthy();
  });

  it('test export function', () => {
    const page = new CPage(mockPageData, {
      materials: mockMaterial,
    });
    expect(page).not.toBeNull();
    expect(page.emitter).not.toBeUndefined();
    expect(
      page.value.componentsTree.value.children[0].material
    ).not.toBeUndefined();
    expect(page.export()).toMatchSnapshot();
  });

  it('test page init with material', () => {
    const page = new CPage(mockPageData, {
      materials: mockMaterial,
    });
    expect(page).not.toBeNull();
    expect(page.emitter).not.toBeUndefined();
    expect(
      page.value.componentsTree.value.children[0].material
    ).not.toBeUndefined();
  });
});
