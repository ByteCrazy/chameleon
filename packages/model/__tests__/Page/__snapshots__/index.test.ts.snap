// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test page model test export function 1`] = `
Object {
  "componentsMeta": Array [
    Object {
      "componentName": "Button",
      "exportName": "Button",
      "package": "demo",
      "version": "1.0.0",
    },
    Object {
      "componentName": "Header",
      "exportName": "Header",
      "package": "demo",
      "version": "1.0.0",
    },
  ],
  "componentsTree": Object {
    "children": Array [
      Object {
        "componentName": "Header",
        "configure": Object {
          "propsSetter": Object {},
        },
        "id": "1",
        "props": Object {
          "age": 12,
          "itemRender": Object {
            "renderType": "COMP",
            "type": "SLOT",
            "value": Array [
              Object {
                "componentName": "Button",
                "configure": Object {
                  "propsSetter": Object {},
                },
                "id": "2",
              },
            ],
          },
          "key": "1",
        },
      },
    ],
    "componentName": "CPage",
  },
  "pageName": "testPage",
  "style": "",
  "version": "1.1.0",
}
`;
